src/privacy_accounting/optimizer/dp_optimizer.py

```diff
@@ -0,0 +1,100 @@
+from math import ceil
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+from src.privacy_accounting.analysis import QueryWithLedger
+from src.privacy_accounting.dp_query import GaussianDPQuery
+from . import WrapperOptimiser
+
+
+class DPOptimiser(WrapperOptimiser):
+
+    def __init__(self,
+                 optimiser,
+                 model,
+                 vector_loss,
+                 dp_sum_query,
+                 num_microbatches=None):
+
+        self.optimiser = optimiser
+        self.model = model
+        self.vector_loss = vector_loss
+        self.dp_sum_query = dp_sum_query
+        self.num_microbatches = num_microbatches
+
+        self._global_parameters = self.dp_sum_query.initial_global_state()
+
+    def fit_batch(self, x: torch.Tensor, y: torch.Tensor):
+
+        loss = self.vector_loss(self.model(x), y)
+
+        param_groups = self.optimiser.param_groups
+
+        sample_state = self.dp_sum_query.initial_sample_state(
+            nest.parameters_to_tensor_groups(param_groups, 'data')
+        )
+        sample_params = self.dp_sum_query.derive_sample_params(self._global_parameters)
+
+        microbatch_size = 1 if self.num_microbatches is None else ceil(loss.shape[0] / self.num_microbatches)
+
+        microbatches_losses = loss.split(microbatch_size, dim=0)
+
+        def process_microbatch(losses, sample_state):
+            self.optimiser.zero_grad()
+            microbatch_loss = losses.mean(dim=0)
+            microbatch_loss.backward(retain_graph=True)
+            record = self.get_grads(param_groups)
+            sample_state = self.dp_sum_query.accumulate_record(sample_params, sample_state, record)
```

Comment:
not 100% sure the methodology behind this - it seems really convoluted to me?

Reply:
This is the way you have to do it. 
1. get the autodiff engine to compute the grad for the specific microbatch loss
2. grab those gradients off the parameters to clip
3. accumulate those gradients with what we have already accumulated. (we cant store this on the parameter itself as the autodiff will overwrite it on the next backwards pass)
4. zero out the grads and more onto the next micro batch
5. When youve done all the microbatches, apply noise and put the gradients back on the parameters so the optimiser knows what to do
6. Step the optimeser

Reply:
could we just comment what the sample params and state are? just to be clear - the sample params are the parameters for the DP mechanism ? state the current accumulated gradients?

Reply:
Yeah will work on terminology throughout, inherited a bit from the TF lib I dont like as well